### PR TITLE
Rule S3366: "this" should not be exposed from constructors

### PR DESCRIPTION
**Pull request by valhristov**

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ThisShouldNotBeExposedFromConstructors.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                            {
+                                return;
+                            }
+
+                            var thisExpression = invocation.ArgumentList.Arguments
+                                .Select(a => a.Expression)
+                                .FirstOrDefault(IsThisExpression);
+
+                            if (thisExpression != null &&
+                                !IsClassMember(invocation.Expression) &&
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ThisShouldNotBeExposedFromConstructors.cs]: Is it useful to do `invocation.Expression.RemoveParenthesis()` instead?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs [hunk 2]
```diff
@@ -45,10 +45,10 @@ public static IEnumerable<INamedTypeSymbol> GetAllNamedTypes(this INamespaceSymb
             }
         }
 
-        public static bool IsObsolete(this ISymbol symbol)
-        {
-            return symbol != null &&
-                symbol.GetAttributes().Any(a => a.AttributeClass.Is(KnownType.System_ObsoleteAttribute));
+        public static bool IsObsolete(this ISymbol symbol)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolHelper.cs]: Was it only line ending change?

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ThisShouldNotBeExposedFromConstructors.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            StaticProperty = this;
+            StaticProperty = ((this));
+
+            Other.StaticMethod(this); // Noncompliant
+            Other.StaticList.Add(this); // Noncompliant
+            Other.StaticProperty = this; // Noncompliant
+            ProgramsStatic.Add(this); // Noncompliant
+            InstanceList.Add(this); // Noncompliant
+            this.InstanceList.Add(this); // Noncompliant
+            InstanceProperty = this;
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ThisShouldNotBeExposedFromConstructors.cs]: Isn't it supposed to be caught?

Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ThisShouldNotBeExposedFromConstructors.cs]: as far as I can understand the rule, you are not supposed to provide potentially uninitialized instance to "other" classes and instances and it is ok if you give it to the same instance...

**Pull request merged**